db/views/reporting_patient_states_per_month_v03.sql

```diff
@@ -0,0 +1,173 @@
+    -- information on assigned facility and parent regions
+    p.assigned_facility_id AS assigned_facility_id,
+    assigned_facility.facility_size as assigned_facility_size,
+    assigned_facility.facility_type as assigned_facility_type,
+    assigned_facility.facility_region_slug as assigned_facility_slug,
+    assigned_facility.facility_region_id as assigned_facility_region_id,
+    assigned_facility.block_slug as assigned_block_slug,
+    assigned_facility.block_region_id as assigned_block_region_id,
+    assigned_facility.district_slug as assigned_district_slug,
+    assigned_facility.district_region_id as assigned_district_region_id,
+    assigned_facility.state_slug as assigned_state_slug,
+    assigned_facility.state_region_id as assigned_state_region_id,
+    assigned_facility.organization_slug as assigned_organization_slug,
+    assigned_facility.organization_region_id as assigned_organization_region_id,
+
+    ------------------------------------------------------------
+    -- information on registration facility and parent regions
+    p.registration_facility_id AS registration_facility_id,
+    registration_facility.facility_size as registration_facility_size,
+    registration_facility.facility_type as registration_facility_type,
+    registration_facility.facility_region_slug as registration_facility_slug,
+    registration_facility.facility_region_id as registration_facility_region_id,
+    registration_facility.block_slug as registration_block_slug,
+    registration_facility.block_region_id as registration_block_region_id,
+    registration_facility.district_slug as registration_district_slug,
+    registration_facility.district_region_id as registration_district_region_id,
+    registration_facility.state_slug as registration_state_slug,
+    registration_facility.state_region_id as registration_state_region_id,
+    registration_facility.organization_slug as registration_organization_slug,
+    registration_facility.organization_region_id as registration_organization_region_id,
+
+    ------------------------------------------------------------
+    -- details of the visit: latest BP, encounter, prescription drug and appointment
+    bps.blood_pressure_id as blood_pressure_id,
+    bps.blood_pressure_facility_id AS bp_facility_id,
+    bps.blood_pressure_recorded_at AS bp_recorded_at,
+    bps.systolic,
+    bps.diastolic,
+
+    visits.encounter_id AS encounter_id,
+    visits.encounter_facility_id AS encounter_facility_id,
+    visits.encounter_recorded_at AS encounter_recorded_at,
+
+    visits.prescription_drug_id AS prescription_drug_id,
+    visits.prescription_drug_facility_id AS prescription_drug_facility_id,
+    visits.prescription_drug_recorded_at AS prescription_drug_recorded_at,
+
+    visits.appointment_id AS appointment_id,
+    visits.appointment_creation_facility_id AS appointment_creation_facility_id,
+    visits.appointment_recorded_at AS appointment_recorded_at,
```

Comment:
From all these facilities, how do we ask "what's the facility that the patient last visited?"

Reply:
A dirty solution would be:
```sql
    CASE
        WHEN visits.visited_at = visits.prescription_drug_recorded_at THEN visits.prescription_drug_facility_id  
        WHEN visits.visited_at = visits.appointment_recorded_at THEN visits.appointment_creation_facility_id    
        ELSE visits.encounter_facility_id
        AS latest_facility_id
```

I was thinking about follow ups where we have to answer if the patient visited a given facility in a month. And that would be:
```sql
'given_facility' IN
array_remove(
        ARRAY[
            (CASE WHEN to_char(encounter_recorded_at, 'YYYY-MM') = month_string THEN encounter_facility_id),
            (CASE WHEN to_char(prescription_drug_recorded_at, 'YYYY-MM') = month_string THEN prescription_drug_facility_id),
            (CASE WHEN to_char(appointment_recorded_at, 'YYYY-MM') = month_string THEN appointment_facility_id)
        ],
    null)
```

Let me think of better solutions, or add these to the view. Good question though!

Reply:
Haha yeah, my initial thought was - "do we need to precompute the "visit_facility_id" in the visits table itself? Maybe 🤔

Reply:
@harimohanraj89 There could be multiple visits in a month across multiple facilities. I'm thinking of adding `visited_facility_ids`, wdyt?